File: CTDL-HK hè (Quá trình 2) 01082024/Bài 2/NGOACDUNG.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
    stack<char> ds; 
    string str; cin >> str;
    if (str.size() == 0) {
        cout << "yes"; return 0;
    }
    for (auto x : str) {
        if (x == '(' || x == '[' || x == '{')
            ds.push(x);
        else {
            if (!ds.empty()) {
                char ch = ds.top(); ds.pop();
                if ((ch == '(' && x != ')') || (ch == '[' && x != ']') || (ch == '{' && x != '}')) {
                    cout << "no"; return 0;
                }
            } else {
                cout << "no"; return 0;
            }
        }
    }

    (!ds.empty()) ? cout << "no" : cout << "yes";
    return 0;
}
// Code của Ty

File: CTDL-HK hè (Quá trình 2) 01082024/Bài 2/NGOACDUNG.py
def is_valid_parentheses(s):
    stack = []
    mapping = {")": "(", "}": "{", "]": "["}
    for char in s:
        if char in mapping.values():
            stack.append(char)
        elif char in mapping.keys():
            if stack == [] or mapping[char] != stack.pop():
                return "no"
        else:
            return "no"
    return "yes" if stack == [] else "no"
# Đoạn code dưới đây sẽ đọc xâu đầu vào và gọi hàm kiểm tra
if __name__ == "__main__":
    input_string = input().strip()
    print(is_valid_parentheses(input_string))
